packages/next-on-pages/src/buildApplication/buildSummary.ts

```diff
@@ -0,0 +1,179 @@
+ * @param staticAssets List of static assets collected during the build.
+ * @param processedVercelOutput Results of processing the Vercel output directory.
+ * @param directoryProcessingResults Results of processing the output directory.
+ */
+export function printBuildSummary(
+	staticAssets: string[],
+	{ vercelOutput }: ProcessedVercelOutput,
+	{
+		functionsMap = new Map(),
+		prerenderedRoutes = new Map(),
+		wasmIdentifiers = new Map(),
+	}: Partial<DirectoryProcessingResults> = {}
+): void {
+	const middlewareFunctions = [...vercelOutput.entries()]
+		.filter(([, { type }]) => type === 'middleware')
+		.map(([path]) => path)
+		.sort((a, b) => a.localeCompare(b));
+	const routeFunctions = new Set(
+		processItemsMap(functionsMap).filter(
+			path => !middlewareFunctions.includes(path.replace(/^\//, ''))
+		)
+	);
+	const prerendered = processItemsMap(prerenderedRoutes);
+	let otherStatic = staticAssets
+		.filter(path => !prerenderedRoutes.has(path))
+		.sort((a, b) => a.localeCompare(b));
+
+	// Move the `/_next` and `/__next` assets to the end of the static assets list
+	otherStatic = [
+		...otherStatic.filter(path => !/^\/__?next/.test(path)),
+		...otherStatic.filter(path => /^\/__?next/.test(path)),
+	];
+
+	let summary = `Build Summary (@cloudflare/next-on-pages v${nextOnPagesVersion})`;
+	summary = addToSummary(summary, 'Middleware Functions', middlewareFunctions);
+	summary = addToSummary(summary, 'Edge Function Routes', [...routeFunctions]);
+	summary = addToSummary(summary, 'Prerendered Routes', prerendered, 20);
+	summary = addToSummary(summary, 'Wasm Files', [...wasmIdentifiers.keys()]);
+	summary = addToSummary(summary, 'Other Static Assets', otherStatic, 5);
+
+	cliLog(summary, { spaced: true, skipDedent: true });
+}
+
+/**
+ * Writes information about the build to a specific directory.
+ *
+ * @param outputDir Output directory.
+ * @param staticAssets List of static assets collected during the build.
+ * @param processedVercelOutput Results of processing the Vercel output directory.
+ * @param directoryProcessingResults Results of processing the output directory.
+ */
+export async function writeBuildInfo(
+	outputDir: string,
+	staticAssets: string[],
+	{ vercelOutput }: ProcessedVercelOutput,
+	{
+		invalidFunctions = new Set(),
+		functionsMap = new Map(),
+		prerenderedRoutes = new Map(),
+		wasmIdentifiers = new Map(),
+	}: Partial<DirectoryProcessingResults> = {}
+): Promise<void> {
+	const currentDir = resolve();
+	const filePath = join(outputDir, 'nop-build-info.json');
+
+	// Change wasm paths to be relative to avoid leaking file system structure or account username.
+	const desensitizedWasmIdentifiers = [...wasmIdentifiers.values()].map(
+		({ originalFileLocation, ...rest }) => ({
+			...rest,
+			originalFileLocation: relative(currentDir, originalFileLocation),
+		})
+	);
+
+	await mkdir(outputDir, { recursive: true });
+	await writeFile(
+		filePath,
+		JSON.stringify(
+			{
+				version: nextOnPagesVersion,
```

Comment:
> Even better, why not include all the [relevant packages versions](https://github.com/cloudflare/next-on-pages/blob/cc6c1140cc9ff656ffb5b9cf728b2e985dcaa139/packages/next-on-pages/src/cli.ts#L226) here?

i didn't want to do that because it would unnecessarily add on more time in the build process to find all the package versions, every single build.

Reply:
mh... yeah good point 😕